File: server/crates/evosim-engine/src/terrain.rs
//! Static rocky terrain: obstacle types + geometry helpers.
//!
//! Substrate port of `src/sim/terrain.ts`. Ships the data shapes,
//! polygon queries, and lobe packing the runtime collision pass
//! consumes. The hand-authored polygons from `terrain-shapes.ts` are
//! deliberately NOT ported here -- they're world data, not substrate,
//! and the engine starts with an empty obstacles list. A future world-
//! config / scene port will populate them.
//!
//! What's NOT ported (kept honest):
//!   - `terrain-shapes.ts` -- hand-authored normalized polygons
//!   - `geology.ts` -- per-world vertex jitter for seeded variation
//!   - the surface modifier maps (`buildTerrainSurfaceMaps`) -- depend on
//!     the wave system, which isn't ported. Wind exposure, wave origin,
//!     shoaling all wait on that.

use serde::{Deserialize, Serialize};

/// A circular sub-collider packed inside an obstacle polygon. The
/// runtime collision sweeps treat lobes like spheres so existing
/// circle-vs-circle logic handles polygon shapes.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ObstacleLobe {
    pub x: f32,
    pub y: f32,
    pub r: f32,
}

/// A polygon vertex (matches the TS `{x, y}` shape).
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct PolygonPoint {
    pub x: f32,
    pub y: f32,
}

/// A static rock. Holds the source polygon (for point-in-polygon
/// evacuation + heightmap construction), its lobe pack (for the hot
/// collision path), and the AABB the spatial index keys off.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Obstacle {
    pub min_x: f32,
    pub min_y: f32,
    pub max_x: f32,
    pub max_y: f32,
    pub lobes: Vec<ObstacleLobe>,
    /// The polygon source. `None` for synthetic / test obstacles built
    /// directly from lobes; production rocks always have one.
    pub polygon: Option<Vec<PolygonPoint>>,
    /// Z-band (rendering layer hint). Reserved for the render port;
    /// not used by collision.
    pub z: f32,
}

/// Lobe geometry constants. Match the TS values exactly so the
/// collision footprint is bit-identical.
const LOBE_R: f32 = 9.0;
const LOBE_PITCH: f32 = 12.0;

/// Standard ray-cast point-in-polygon. Translated verbatim from
/// `terrain.ts:74`. Used by the lobe packer + obstacle evacuation.
pub fn point_in_polygon(x: f32, y: f32, poly: &[PolygonPoint]) -> bool {
    if poly.len() < 3 {
        return false;
    }
    let mut inside = false;
    let n = poly.len();
    let mut j = n - 1;
    for i in 0..n {
        let xi = poly[i].x;
        let yi = poly[i].y;
        let xj = poly[j].x;
        let yj = poly[j].y;
        let intersect =
            ((yi > y) != (yj > y)) && (x < (xj - xi) * (y - yi) / (yj - yi + 1e-12) + xi);
        if intersect {
            inside = !inside;
        }
        j = i;
    }
    inside
}

/// Pack a polygon's interior with circular collision lobes on a
/// `LOBE_PITCH` grid, kept where the grid point falls inside the
/// polygon. Sliver polygons that miss every grid point get small
/// vertex-anchored lobes as a fallback so collision still has
/// something to push off.
pub fn lobes_from_terrain_polygon(polygon: &[PolygonPoint]) -> Vec<ObstacleLobe> {
    if polygon.is_empty() {
        return Vec::new();
    }
    let mut min_x = f32::INFINITY;
    let mut min_y = f32::INFINITY;
    let mut max_x = f32::NEG_INFINITY;
    let mut max_y = f32::NEG_INFINITY;
    for v in polygon {
        if v.x < min_x {
            min_x = v.x;
        }
        if v.y < min_y {
            min_y = v.y;
        }
        if v.x > max_x {
            max_x = v.x;
        }
        if v.y > max_y {
            max_y = v.y;
        }
    }
    let mut lobes = Vec::new();
    let mut y = min_y + LOBE_PITCH * 0.5;
    while y <= max_y {
        let mut x = min_x + LOBE_PITCH * 0.5;
        while x <= max_x {
            if point_in_polygon(x, y, polygon) {
                lobes.push(ObstacleLobe { x, y, r: LOBE_R });
            }
            x += LOBE_PITCH;
        }
        y += LOBE_PITCH;
    }
    if lobes.is_empty() {
        for v in polygon {
            lobes.push(ObstacleLobe {
                x: v.x,
                y: v.y,
                r: LOBE_R * 0.6,
            });
        }
    }
    lobes
}

/// Build an Obstacle from a polygon (lobe-packs it, computes the
/// AABB from the polygon's vertices). The polygon is stored on the
/// obstacle for the evacuation path.
pub fn make_obstacle_from_polygon(polygon: Vec<PolygonPoint>) -> Option<Obstacle> {
    if polygon.len() < 3 {
        return None;
    }
    let lobes = lobes_from_terrain_polygon(&polygon);
    // AABB from the polygon vertices (not the lobes -- a vertex may
    // sit outside the lobe pack but is still part of the silhouette).
    let mut min_x = f32::INFINITY;
    let mut min_y = f32::INFINITY;
    let mut max_x = f32::NEG_INFINITY;
    let mut max_y = f32::NEG_INFINITY;
    for v in &polygon {
        if v.x < min_x {
            min_x = v.x;
        }
        if v.y < min_y {
            min_y = v.y;
        }
        if v.x > max_x {
            max_x = v.x;
        }
        if v.y > max_y {
            max_y = v.y;
        }
    }
    Some(Obstacle {
        min_x,
        min_y,
        max_x,
        max_y,
        lobes,
        polygon: Some(polygon),
        z: 0.0,
    })
}

/// Per-column heightmap: topmost rock y for every integer x in
/// `0..width`. Columns with no rock report `f32::INFINITY` (the
/// "open water" sentinel).
pub fn build_terrain_heightmap(obstacles: &[Obstacle], width: f32) -> Vec<f32> {
    let cols = (width.floor() as usize).max(1);
    let mut heightmap = vec![f32::INFINITY; cols];
    for ob in obstacles {
        let Some(poly) = &ob.polygon else { continue };
        let x0 = (ob.min_x.floor() as i32).max(0) as usize;
        let x1 = (ob.max_x.ceil() as i32).min(cols as i32 - 1) as usize;
        let col_lo = x0;
        let col_hi = x1.min(cols.saturating_sub(1));
        for (col, slot) in heightmap
            .iter_mut()
            .enumerate()
            .take(col_hi + 1)
            .skip(col_lo)
        {
            let x = col as f32;
            let mut top_y = f32::INFINITY;
            let n = poly.len();
            let mut j = n - 1;
            for i in 0..n {
                let xi = poly[i].x;
                let yi = poly[i].y;
                let xj = poly[j].x;
                let yj = poly[j].y;
                let in_x_range =
                    (xi <= x && xj >= x) || (xj <= x && xi >= x);
                if in_x_range {
                    if (xi - xj).abs() < 1e-9 {
                        // Vertical edge: every y on it is a candidate.
                        if yi < top_y {
                            top_y = yi;
                        }
                        if yj < top_y {
                            top_y = yj;
                        }
                    } else {
                        let t = (x - xi) / (xj - xi);
                        let y = yi + (yj - yi) * t;
                        if y < top_y {
                            top_y = y;
                        }
                    }
                }
                j = i;
            }
            if top_y < *slot {
                *slot = top_y;
            }
        }
    }
    heightmap
}

/// True if a candidate body of radius `body_r` at `(x, y)` would
/// overlap any obstacle's lobe pack. Used at founder spawn so cells
/// don't materialise inside rock.
pub fn founder_terrain_blocked(obstacles: &[Obstacle], x: f32, y: f32, body_r: f32) -> bool {
    for ob in obstacles {
        if x + body_r < ob.min_x || x - body_r > ob.max_x {
            continue;
        }
        if y + body_r < ob.min_y || y - body_r > ob.max_y {
            continue;
        }
        for l in &ob.lobes {
            let dx = x - l.x;
            let dy = y - l.y;
            let min_dist = body_r + l.r;
            if dx * dx + dy * dy < min_dist * min_dist {
                return true;
            }
        }
    }
    false
}

/// Closest point on a polygon's boundary to `(qx, qy)`. Walks every
/// edge, projects the query onto each segment, returns the
/// projection nearest in Euclidean distance.
pub fn nearest_polygon_edge_point(qx: f32, qy: f32, poly: &[PolygonPoint]) -> (f32, f32) {
    if poly.is_empty() {
        return (qx, qy);
    }
    let mut best_x = poly[0].x;
    let mut best_y = poly[0].y;
    let mut best_d = f32::INFINITY;
    let n = poly.len();
    let mut j = n - 1;
    for i in 0..n {
        let ax = poly[j].x;
        let ay = poly[j].y;
        let bx = poly[i].x;
        let by = poly[i].y;
        let ex = bx - ax;
        let ey = by - ay;
        let len_sq = ex * ex + ey * ey;
        let mut t = 0.0_f32;
        if len_sq > 1e-12 {
            t = ((qx - ax) * ex + (qy - ay) * ey) / len_sq;
            t = t.clamp(0.0, 1.0);
        }
        let px = ax + ex * t;
        let py = ay + ey * t;
        let dx = px - qx;
        let dy = py - qy;
        let d = dx * dx + dy * dy;
        if d < best_d {
            best_d = d;
            best_x = px;
            best_y = py;
        }
        j = i;
    }
    (best_x, best_y)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn square(min: f32, max: f32) -> Vec<PolygonPoint> {
        vec![
            PolygonPoint { x: min, y: min },
            PolygonPoint { x: max, y: min },
            PolygonPoint { x: max, y: max },
            PolygonPoint { x: min, y: max },
        ]
    }

    #[test]
    fn point_in_polygon_basic_square() {
        let sq = square(0.0, 10.0);
        assert!(point_in_polygon(5.0, 5.0, &sq));
        assert!(!point_in_polygon(20.0, 5.0, &sq));
        assert!(!point_in_polygon(-1.0, 5.0, &sq));
        assert!(!point_in_polygon(5.0, -1.0, &sq));
    }

    #[test]
    fn lobes_pack_inside_polygon() {
        let sq = square(0.0, 60.0);
        let lobes = lobes_from_terrain_polygon(&sq);
        assert!(!lobes.is_empty());
        for l in &lobes {
            assert_eq!(l.r, LOBE_R);
            assert!(l.x > 0.0 && l.x < 60.0);
            assert!(l.y > 0.0 && l.y < 60.0);
            assert!(point_in_polygon(l.x, l.y, &sq));
        }
    }

    #[test]
    fn make_obstacle_aabb_matches_polygon() {
        let poly = vec![
            PolygonPoint { x: 10.0, y: 20.0 },
            PolygonPoint { x: 30.0, y: 20.0 },
            PolygonPoint { x: 30.0, y: 50.0 },
            PolygonPoint { x: 10.0, y: 50.0 },
        ];
        let ob = make_obstacle_from_polygon(poly).unwrap();
        assert_eq!(ob.min_x, 10.0);
        assert_eq!(ob.min_y, 20.0);
        assert_eq!(ob.max_x, 30.0);
        assert_eq!(ob.max_y, 50.0);
        assert!(ob.polygon.is_some());
    }

    #[test]
    fn make_obstacle_rejects_degenerate() {
        assert!(make_obstacle_from_polygon(vec![]).is_none());
        assert!(make_obstacle_from_polygon(vec![PolygonPoint { x: 0.0, y: 0.0 }]).is_none());
    }

    #[test]
    fn heightmap_reports_top_of_rock() {
        // Polygon spans columns 10..30, top y = 20.
        let poly = vec![
            PolygonPoint { x: 10.0, y: 20.0 },
            PolygonPoint { x: 30.0, y: 20.0 },
            PolygonPoint { x: 30.0, y: 50.0 },
            PolygonPoint { x: 10.0, y: 50.0 },
        ];
        let ob = make_obstacle_from_polygon(poly).unwrap();
        let hm = build_terrain_heightmap(&[ob], 60.0);
        // Columns under the rock report y=20.
        assert!((hm[15] - 20.0).abs() < 1e-3);
        assert!((hm[29] - 20.0).abs() < 1e-3);
        // Columns outside report Infinity.
        assert!(hm[0].is_infinite());
        assert!(hm[40].is_infinite());
    }

    #[test]
    fn founder_blocked_inside_rock() {
        let poly = square(0.0, 60.0);
        let ob = make_obstacle_from_polygon(poly).unwrap();
        assert!(founder_terrain_blocked(std::slice::from_ref(&ob), 30.0, 30.0, 5.0));
        assert!(!founder_terrain_blocked(&[ob], 200.0, 200.0, 5.0));
    }

    #[test]
    fn nearest_edge_clamps_to_segment() {
        let sq = square(0.0, 10.0);
        // Query inside center, nearest point should be on one of the
        // four edges at distance 5.
        let (x, y) = nearest_polygon_edge_point(5.0, 5.0, &sq);
        let d = ((x - 5.0).hypot(y - 5.0)).abs();
        assert!((d - 5.0).abs() < 1e-3, "expected dist=5, got {d}");
    }

    #[test]
    fn empty_obstacles_no_blocked() {
        // Founder spawn on a terrain-less world should never reject.
        assert!(!founder_terrain_blocked(&[], 0.0, 0.0, 5.0));
        assert!(!founder_terrain_blocked(&[], 500.0, 500.0, 5.0));
    }
}

File: src/sim/terrain.ts
// Static rocky terrain: geometry packing, derived per-column maps, and
// the cheap point/disc terrain queries. All pure or world-mutating-only
// helpers -- no RNG, no module-level state. generateObstacles in sim.ts
// orchestrates these (it owns the vent wiring + the SURFACE_Y_FRAC the
// surface maps key off), but the heavy geometry math lives here.
//
// Geometry comes from hand-authored normalized polygons in
// ./terrain-shapes.ts, scaled to the world and lobe-packed so the
// existing circle-vs-circle obstacle collision handles them.

import type { World, Obstacle, ObstacleLobe } from "./core";

// Push a terrain polygon as an Obstacle with lobes packed inside.
// Centralizes the bounding-box / color / z bookkeeping the loop in
// generateObstacles used to repeat per rock.
export function pushTerrainPolygon(world: World, polygon: { x: number; y: number }[], color: string): void {
  if (polygon.length < 3) return;
  const lobes = lobesFromTerrainPolygon(polygon);
  const ob = makeObstacleFromLobes(lobes, color);
  ob.polygon = polygon;
  ob.z = 0;
  for (const v of polygon) {
    if (v.x < ob.minX) ob.minX = v.x;
    if (v.y < ob.minY) ob.minY = v.y;
    if (v.x > ob.maxX) ob.maxX = v.x;
    if (v.y > ob.maxY) ob.maxY = v.y;
  }
  world.obstacles.push(ob);
}

// Pack a polygon's interior with collision lobes. Strategy: rasterize
// a coarse interior grid (one sample every LOBE_PITCH px), keep any
// grid point that's >=LOBE_R inside the polygon, drop a lobe there.
// Yields a "+"-pattern fill that conservatively underapproximates
// the polygon -- particles never tunnel through, but a particle can
// graze a polygon corner without contact. Acceptable for terrain.
function lobesFromTerrainPolygon(polygon: { x: number; y: number }[]): ObstacleLobe[] {
  // Coarse-ish lobe packing -- evacuateRocks is the definitive
  // "inside rock" check via polygon point-in-polygon, so the lobes
  // only need to handle the cheap radial-pushback case for bodies
  // skimming the boundary. Anything that gets all the way inside a
  // polygon is dealt with by the evacuation pass.
  const LOBE_R = 9;
  const LOBE_PITCH = 12;
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  for (const v of polygon) {
    if (v.x < minX) minX = v.x;
    if (v.y < minY) minY = v.y;
    if (v.x > maxX) maxX = v.x;
    if (v.y > maxY) maxY = v.y;
  }
  const lobes: ObstacleLobe[] = [];
  // Sample grid. Offset by half-pitch so two adjacent thin polygons
  // (e.g. cave ceiling + lip) don't share a sample row and end up
  // with concentric lobes redundantly covering the boundary.
  for (let y = minY + LOBE_PITCH * 0.5; y <= maxY; y += LOBE_PITCH) {
    for (let x = minX + LOBE_PITCH * 0.5; x <= maxX; x += LOBE_PITCH) {
      if (pointInPolygon(x, y, polygon)) {
        lobes.push({ x, y, r: LOBE_R });
      }
    }
  }
  // Edge fallback: a sliver polygon (e.g. very thin lip) might fit no
  // interior sample. Drop a small lobe at each vertex so collision
  // still has SOMETHING to push off.
  if (lobes.length === 0) {
    for (const v of polygon) lobes.push({ x: v.x, y: v.y, r: LOBE_R * 0.6 });
  }
  return lobes;
}

// Standard ray-cast point-in-polygon. Used by the lobe packer above and
// by the obstacle-evacuation pass in sim.ts.
export function pointInPolygon(x: number, y: number, poly: { x: number; y: number }[]): boolean {
  let inside = false;
  for (let i = 0, j = poly.length - 1; i < poly.length; j = i++) {
    const xi = poly[i].x, yi = poly[i].y;
    const xj = poly[j].x, yj = poly[j].y;
    const intersect = ((yi > y) !== (yj > y)) &&
      (x < (xj - xi) * (y - yi) / (yj - yi + 1e-12) + xi);
    if (intersect) inside = !inside;
  }
  return inside;
}

function makeObstacleFromLobes(lobes: ObstacleLobe[], color: string): Obstacle {
  let minX = Infinity, minY = Infinity, maxX = -Infinity, maxY = -Infinity;
  for (const l of lobes) {
    if (l.x - l.r < minX) minX = l.x - l.r;
    if (l.y - l.r < minY) minY = l.y - l.r;
    if (l.x + l.r > maxX) maxX = l.x + l.r;
    if (l.y + l.r > maxY) maxY = l.y + l.r;
  }
  return { minX, minY, maxX, maxY, lobes, color, z: 0 };
}

// Per-column heightmap: topmost rock y per integer x across every
// obstacle. Used as a cheap early-reject in founderTerrainBlocked /
// topTerrainYAtColumn. founderTerrainBlocked falls through to a
// per-obstacle lobe check anyway, so this is purely a fast-path hint
// -- a column that has rock high up but clear water below still gets
// the heightmap top, and the lobe sweep refines the actual
// overlap. Initialized to +Inf so columns with no rock report
// "open water" via the existing semantics.
export function buildTerrainHeightmap(obstacles: Obstacle[], width: number): Float32Array {
  const heightmap = new Float32Array(Math.max(1, Math.floor(width)));
  heightmap.fill(Number.POSITIVE_INFINITY);
  for (const ob of obstacles) {
    if (!ob.polygon) continue;
    const x0 = Math.max(0, Math.floor(ob.minX));
    const x1 = Math.min(heightmap.length - 1, Math.ceil(ob.maxX));
    for (let x = x0; x <= x1; x++) {
      // Top y at this column = smallest y the polygon covers (y down).
      // Done by ray-casting at integer x: find min y where the polygon
      // interior begins. Cheap (one ray per column, <50 vertices).
      let topY = Number.POSITIVE_INFINITY;
      // Walk vertical edges of the polygon; the smallest y where the
      // ray (x = const) crosses an edge bounds the polygon span. Then
      // we also accept any vertex with the matching x range -- a flat
      // top edge contributes only its endpoints to crossings.
      const poly = ob.polygon;
      for (let i = 0, j = poly.length - 1; i < poly.length; j = i++) {
        const xi = poly[i].x, yi = poly[i].y;
        const xj = poly[j].x, yj = poly[j].y;
        if ((xi <= x && xj >= x) || (xj <= x && xi >= x)) {
          if (xi === xj) {
            // Vertical edge: every y on it is a potential top.
            if (yi < topY) topY = yi;
            if (yj < topY) topY = yj;
          } else {
            const t = (x - xi) / (xj - xi);
            const y = yi + (yj - yi) * t;
            if (y < topY) topY = y;
          }
        }
      }
      if (topY < heightmap[x]) heightmap[x] = topY;
    }
  }
  return heightmap;
}

// Per-column surface-modifier maps derived from the heightmap: how the
// wavy water surface responds to rocks that breach (or nearly breach)
// the still water level at stillSurfaceY. All three are direction-paired
// (from-left when wind blows right, from-right otherwise) and consumed
// by the surface helpers in ./environment.
export interface TerrainSurfaceMaps {
  windExposureFromLeft: Float32Array;
  windExposureFromRight: Float32Array;
  waveOriginFromLeft: Float32Array;
  waveOriginFromRight: Float32Array;
  shoalFromLeft: Float32Array;
  shoalFromRight: Float32Array;
}

export function buildTerrainSurfaceMaps(heightmap: Float32Array, stillSurfaceY: number): TerrainSurfaceMaps {
  // Wind exposure maps. cliffHeight[x] = how far above the still
  // water surface the rock extends at column x (0 if no rock pokes
  // out of the water). A tall cliff casts a wind shadow downwind
  // proportional to its height; exposure ramps from 0 right behind
  // the cliff back up to 1 over SHELTER_DIST_MAX pixels.
  const cliffHeight = new Float32Array(heightmap.length);
  for (let x = 0; x < heightmap.length; x++) {
    const top = heightmap[x];
    if (top !== Number.POSITIVE_INFINITY && top < stillSurfaceY) {
      cliffHeight[x] = stillSurfaceY - top;
    }
  }
  // Shelter distance scales linearly with cliff height so a tiny
  // bump shelters only a few px while a tall cliff shelters far
  // downwind.
  const SHELTER_DIST_PER_HEIGHT = 2;
  const SHELTER_DIST_MAX = 180;
  const buildExposureMap = (fromLeft: boolean): Float32Array => {
    const m = new Float32Array(heightmap.length);
    m.fill(1);
    // For each column, scan UPWIND looking for the tallest cliff
    // within shelter range. The shadow decays linearly with distance.
    const step = fromLeft ? -1 : 1;
    for (let x = 0; x < heightmap.length; x++) {
      // If rock fills this column in the air band, exposure 0.
      if (cliffHeight[x] > 0 && heightmap[x] <= 1) { m[x] = 0; continue; }
      let bestShadow = 0;
      const maxScan = SHELTER_DIST_MAX;
      for (let d = 1; d <= maxScan; d++) {
        const ux = x + step * d;
        if (ux < 0 || ux >= heightmap.length) break;
        const ch = cliffHeight[ux];
        if (ch <= 0) continue;
        const reach = Math.min(SHELTER_DIST_MAX, ch * SHELTER_DIST_PER_HEIGHT);
        if (d > reach) continue;
        // Shelter strength: cliff height normalized vs surfaceY, with
        // linear distance falloff.
        const strength = Math.min(1, ch / stillSurfaceY) * (1 - d / reach);
        if (strength > bestShadow) bestShadow = strength;
      }
      m[x] = Math.max(0, 1 - bestShadow);
    }
    return m;
  };

  // Wave-origin maps: the spatial phase reference for each column. As the
  // wave train sweeps in from the upwind edge, every surface-breaching
  // rock resets the origin to its own (downwind) edge, so leeward water
  // re-forms its own waves from the rock rather than continuing the
  // windward phase. Without this the surface is one global sine and a
  // crest appears to translate straight through a rock.
  const buildWaveOrigin = (fromLeft: boolean): Float32Array => {
    const m = new Float32Array(cliffHeight.length);
    if (fromLeft) {
      let origin = 0;
      for (let x = 0; x < cliffHeight.length; x++) {
        if (cliffHeight[x] > 0) origin = x; // tracks to the rock's far edge
        m[x] = origin;
      }
    } else {
      let origin = cliffHeight.length - 1;
      for (let x = cliffHeight.length - 1; x >= 0; x--) {
        if (cliffHeight[x] > 0) origin = x;
        m[x] = origin;
      }
    }
    return m;
  };

  // Shoaling maps: a wave bound for a surface-breaching rock piles up as
  // it reaches the shallows in front of it. For each open-water column,
  // find the nearest breaching rock DOWNWIND within SHOAL_APRON_PX and
  // boost amplitude, strongest right at the rock face -- so the wave
  // steepens into the rock and then breaks (the rock columns are glassy
  // via exposure 0). Direction-paired: downwind is +x for rightward wind.
  const SHOAL_APRON_PX = 55;
  const SHOAL_GAIN = 1.1; // up to ~2.1x amplitude at the face
  const buildShoal = (fromLeft: boolean): Float32Array => {
    const m = new Float32Array(cliffHeight.length);
    m.fill(1);
    const step = fromLeft ? 1 : -1; // scan downwind for the rock
    for (let x = 0; x < cliffHeight.length; x++) {
      if (cliffHeight[x] > 0) continue; // on the rock: no shoaling (glassy)
      for (let d = 1; d <= SHOAL_APRON_PX; d++) {
        const ux = x + step * d;
        if (ux < 0 || ux >= cliffHeight.length) break;
        if (cliffHeight[ux] > 0) { // rock d px downwind -> shoal, peak at the face
          m[x] = 1 + SHOAL_GAIN * (1 - (d - 1) / SHOAL_APRON_PX);
          break;
        }
      }
    }
    return m;
  };

  return {
    windExposureFromLeft: buildExposureMap(true),
    windExposureFromRight: buildExposureMap(false),
    waveOriginFromLeft: buildWaveOrigin(true),
    waveOriginFromRight: buildWaveOrigin(false),
    shoalFromLeft: buildShoal(true),
    shoalFromRight: buildShoal(false),
  };
}

// True if a candidate body of radius `bodyR` centered at (x, y) would
// overlap any rock. Conservative: tests the candidate disc against
// every obstacle's lobes, which is what the runtime collision pass
// would push out anyway. Used at founder spawn so cells don't enter
// inside the cave, under an outcropping, or stuck in the seafloor.
export function founderTerrainBlocked(world: World, x: number, y: number, bodyR: number): boolean {
  // Per-obstacle lobe test against every rock. With ~5 obstacles and
  // ~30 lobes each this is ~150 ops per attempt -- fine. (The
  // heightmap fast-path that used to short-circuit here only worked
  // for a monotonic seafloor; the hand-authored layout has top-down
  // cliffs that don't fit that assumption, so we rely solely on the
  // lobe sweep now.)
  for (const ob of world.obstacles) {
    if (x + bodyR < ob.minX || x - bodyR > ob.maxX) continue;
    if (y + bodyR < ob.minY || y - bodyR > ob.maxY) continue;
    const lobes = ob.lobes;
    for (let j = 0; j < lobes.length; j++) {
      const l = lobes[j];
      const dx = x - l.x;
      const dy = y - l.y;
      const minDist = bodyR + l.r;
      if (dx * dx + dy * dy < minDist * minDist) return true;
    }
  }
  return false;
}

// Topmost rock surface at column x. Looks up world.terrainHeightmap;
// returns +Inf for columns with no rock (or worlds without terrain).
export function topTerrainYAtColumn(world: World, x: number): number {
  const hm = world.terrainHeightmap;
  if (!hm || hm.length === 0) return Infinity;
  const ix = Math.max(0, Math.min(hm.length - 1, Math.floor(x)));
  return hm[ix];
}
